include/jemalloc/internal/tcache_inlines.h

```diff
@@ -33,68 +33,80 @@ tcache_enabled_set(tsd_t *tsd, bool enabled) {
 }
 
 static inline unsigned
-tcache_nhbins_get(tcache_t *tcache) {
-	assert(tcache != NULL);
-	assert(tcache->tcache_nhbins <= TCACHE_NBINS_MAX);
-	return tcache->tcache_nhbins;
+tcache_nbins_get(tcache_slow_t *tcache_slow) {
+	assert(tcache_slow != NULL);
+	assert(tcache_slow->tcache_nbins <= TCACHE_NBINS_MAX);
+	return tcache_slow->tcache_nbins;
 }
 
 static inline size_t
-tcache_max_get(tcache_t *tcache) {
-	assert(tcache != NULL);
-	assert(tcache->tcache_max <= TCACHE_MAXCLASS_LIMIT);
-	return tcache->tcache_max;
+tcache_max_get(tcache_slow_t *tcache_slow) {
+	assert(tcache_slow != NULL);
+	size_t tcache_max = sz_index2size(tcache_nbins_get(tcache_slow) - 1);
+	assert(tcache_max <= TCACHE_MAXCLASS_LIMIT);
+	return tcache_max;
 }
 
 static inline void
-tcache_max_and_nhbins_set(tcache_t *tcache, size_t tcache_max) {
-	assert(tcache != NULL);
+tcache_max_set(tcache_slow_t *tcache_slow, size_t tcache_max) {
+	assert(tcache_slow != NULL);
 	assert(tcache_max <= TCACHE_MAXCLASS_LIMIT);
-	tcache->tcache_max = tcache_max;
-	tcache->tcache_nhbins = sz_size2index(tcache_max) + 1;
+	tcache_slow->tcache_nbins = sz_size2index(tcache_max) + 1;
+}
+
+static inline void
+tcache_bin_settings_backup(tcache_t *tcache,
+    cache_bin_info_t *tcache_bin_info) {
```

Comment:
Use `cache_bin_info_t tcache_bin_info[TCACHE_NBINS_MAX]` instead of a plain pointer. That way compiler can warn us if the size is not as expected.